dojson: fixes from validation

ioannistsanaktsidis opened the pull request:
Signed-off-by: Ioannis Tsanaktsidis <ioannis.tsanaktsidis@cern.ch>

ioannistsanaktsidis commented:
This PR fixes mistakes that came out from validating against jsoncschema. We should check the changes that I did and then we can merge it.

kaplun commented:
Hi @ioannistsanaktsidis, I have tried your branch locally and it didn't crash. Can you try to rebase it and push it again?

ioannistsanaktsidis commented:
Hi @kaplun  I have just rebased and pushed the branch again. The branch wasn't crashing also for me. Although there was a problem when I was trying to upload records. Some of them were crashing and not uploaded.(e.g try the demo records). I guess it was a mapping problem.

kaplun commented:
Mmh. Now it is a more useful crash :-) It actually crashes in the workflows test. @jalavik can you help us debugging it?

jalavik commented:
There are PEP8 errors as well causing the failure.

Regarding the workflow crash, it looks to be related to the model changes related to arxiv ids:

```
File "/home/travis/build/inspirehep/inspire-next/inspire/modules/oaiharvester/tasks/arxiv.py", line 66, in get_arxiv_id_from_record
                           if not arxiv_id.lower().startswith("oai:arxiv") and not \
AttributeError: 'bool' object has no attribute 'lower'
```
Somehow arxiv_id is bool now.

kaplun commented:
> There are PEP8 errors as well causing the failure.

For that I guess we can plug @jirikuncar's kwalitee with GitHub.

ioannistsanaktsidis commented:
I think this is happening because I changed the primary value of arxiv to
be bool, because of Json schema.
@hep.over('report_number', '^037..')#@utils.for_each_value
#@utils.filter_valuesdef report_number(self, key, value): """Source of
Acquisition.""" def get_value(value): return { 'primary': bool(value.get('a'
)), 'arxiv_category': value.get('c'), 'source': value.get('9')

} I did that because of Json schema, that requires this:
"report_number": {
            "uniqueItems": true,
            "items": {
                "type": "object",
                "properties": {
                    "source": {
                        "type": "string"
                    },
                    "primary": {
                        *"type": "boolean"*
                    },
                    "value": {
                        "type": "string"
                    }
                }
            },
            "type": "array"
        },

2015-09-21 10:55 GMT+02:00 Jan Åge Lavik <notifications@github.com>:

> There are PEP8 errors as well causing the failure.
>
> Regarding the workflow crash, it looks to be related to the model changes
> related to arxiv ids:
>
> File "/home/travis/build/inspirehep/inspire-next/inspire/modules/oaiharvester/tasks/arxiv.py", line 66, in get_arxiv_id_from_record
>                            if not arxiv_id.lower().startswith("oai:arxiv") and not \
> AttributeError: 'bool' object has no attribute 'lower'
>
> Somehow arxiv_id is bool now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inspirehep/inspire-next/pull/362#issuecomment-141915125>
> .
>

jalavik commented:
> For that I guess we can plug @jirikuncar's kwalitee with GitHub.

This is not strictly needed as the PEP8 errors are already reported in Travis thanks to the pytest integration. Check the output above the exception info.

kaplun commented on inspire/dojson/hep/fields/bd01x09x.py:
```diff
@@ -95,15 +95,23 @@ def system_control_number2marc(self, key, value):
-    }
+    def get_value(value):
+        return {
+            'primary': bool(value.get('a')),
```
Uh! Actually primary is whether the reportnumber is primary (hence `037`) or not (hence `088` tag). $a contains instead what should into the `value` JSON key. So in the context of `037` you can always put `True` in the `primary` field.

jalavik replied:
Yeah, and can you `git grep` for `report_number.primary` lookups and change them to `report_number.value`. Same approach should in fact to applied to all keys that change their value or lookup key in this PR.

ioannistsanaktsidis replied:
I amended the commit with the latest changes. I have also to grep  and
changes the values wherever they are used so as the code not to crash. Feel
free to check the latest changes about the report_number and the
arxiv_eprints.

2015-09-21 12:54 GMT+02:00 Jan Åge Lavik <notifications@github.com>:

> In inspire/dojson/hep/fields/bd01x09x.py
> <https://github.com/inspirehep/inspire-next/pull/362#discussion_r39958960>
> :
>
> >  def report_number(self, key, value):
> >      """Source of Acquisition."""
> > -    return {
> > -        'primary': value.get('a', value.get('z')),
> > -        'arxiv_category': value.get('c'),
> > -        'source': value.get('9')
> > -    }
> > +    def get_value(value):
> > +        return {
> > +            'primary': bool(value.get('a')),
>
> Yeah, and can you git grep for report_number.primary lookups and change
> them to report_number.value. Same approach should in fact to applied to
> all keys that change their value or lookup key in this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inspirehep/inspire-next/pull/362/files#r39958960>.
>

jalavik commented:
@ioannistsanaktsidis Cool. Thanks dude. Can you also take a look at the PEP8 errors in the travis output? https://travis-ci.org/inspirehep/inspire-next/builds/81373507

ioannistsanaktsidis commented:
@jalavik for now we are passing all tests! I am going to take care every time I update the PR. Cheers!

jalavik commented:
@ioannistsanaktsidis Will now make a conversion from json to marc, then add some tests and this PR will be good to go.

kaplun commented:
:+1:

ioannistsanaktsidis commented:
The test is implemented and pushed. I have some minor questions and that need to be solved and then the branch is ready for merging.

ioannistsanaktsidis commented:
After the latest push and the changes that I made, I detected that there is an error caused when someone tries to create records. This error is caused due to the update of publication_info field(773). It is caused because I updated the 'year' to be an integer. You can check in the travis the full trace back. Let me say that I also updated the mapping in the hep.json so as the year to be integer. So I guess we have to change sth else too so as to be fixed. :)

jalavik commented:
@ioannistsanaktsidis The way testing is done has been updated so you can:

* Rebase with latest master
* Remove WIP from commit log message
* Remove the comments we talked about causing the PEP8 issue

Thanks!

ioannistsanaktsidis commented:
@jalavik the bullet points are now fixed. I need to check some staff on Monday and then we can talk about the merging. Cheers.

ioannistsanaktsidis commented:
@jalavik  Evth is up to date! We can when you have some time and check evth one more time. Than If evth is of we can merge the PR. Cheers!

jalavik commented on inspire/modules/workflows/tasks/matching.py:
```diff
@@ -248,4 +248,5 @@ def _update_old_object(obj, eng):
     """Temporary measure to enable sorting by primary category."""
     record = get_record_from_obj(obj, eng)
-    obj.uri = record.get("report_number.arxiv_category", [""])[0]
+    obj.uri = record['arxiv_eprints'][0]['categories'][0]
```
I think this could still be accessed like `record.get("arxiv_eprints.categories". [""])[0]` (as `record` is a `Record` object at this point. If you can try it it would be great. Then you can also remove the commented line below.

ioannistsanaktsidis replied:
Hey Jan ,
I tried it and it was failing the test, so thats why I changed and i left
the comment so as to check it together. I will double check tomorrow and I
will inform you. Cheers!
Στις 28 Σεπ 2015 5:57 μ.μ., ο χρήστης "Jan Åge Lavik" <
notifications@github.com> έγραψε:

> In inspire/modules/workflows/tasks/matching.py
> <https://github.com/inspirehep/inspire-next/pull/362#discussion_r40569265>
> :
>
> > @@ -248,4 +248,5 @@ def _update_old_object(obj, eng):
> >  def arxiv_set_category_field(obj, eng):
> >      """Temporary measure to enable sorting by primary category."""
> >      record = get_record_from_obj(obj, eng)
> > -    obj.uri = record.get("report_number.arxiv_category", [""])[0]
> > +    obj.uri = record['arxiv_eprints'][0]['categories'][0]
>
> I think this could still be accessed like record.get("arxiv_eprints.categories".
> [""])[0] (as record is a Record object at this point. If you can try it
> it would be great. Then you can also remove the commented line below.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inspirehep/inspire-next/pull/362/files#r40569265>.
>

jalavik replied:
OK. Cool, man!

ioannistsanaktsidis replied:
I tried this and it works, so we can keep this version in the end:
record.get("arxiv_eprints.categories".
[""])[0]

jalavik commented on inspire/testsuite/test_hep.py:
```diff
@@ -0,0 +1,200 @@
+class HepRecordsTests(InvenioTestCase):
+    def setUp(self):
+        self.marcxml_to_json = list(processors.convert_marcxml(
+            open('inspire/testsuite/fixtures/test_hep_record.xml',
```
Generally, it is safer to use the pkg_resources (like https://github.com/inveniosoftware/invenio-records/blob/master/tests/test_record.py#L40-L42):

```python
marcxml = pkg_resources.resource_string('inspire.testsuite', os.path.join('fixtures', 'test_hep_record.xml'))
```

jalavik commented:
@ioannistsanaktsidis Well done! It should be good to go now after the minor comments I just put ;-) Thanks!

ioannistsanaktsidis commented:
@jalavik everything up to date !

jalavik commented:
@ioannistsanaktsidis Thanks. I will do a final test locally and merge.

ioannistsanaktsidis commented:
Cool man!


2015-09-29 11:45 GMT+02:00 Jan Åge Lavik <notifications@github.com>:

> @ioannistsanaktsidis <https://github.com/ioannistsanaktsidis> Thanks. I
> will do a final test locally and merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inspirehep/inspire-next/pull/362#issuecomment-144008147>
> .
>
